File: 08_StacksAndQueues/Queues/Queues PY/Queues.py
class Queue:
    def __init__(self):
        self.front = -1
        self.rear = -1
        self.size = 7
        self.arr = [None] * self.size

    def is_full(self):
        return self.front == 0 and self.rear == self.size - 1

    def is_empty(self):
        return self.front == -1

    def enqueue(self, val):
        if self.is_full():
            print("Full Queue")
            return
        if self.front == -1:
            self.front = 0
        self.rear += 1
        self.arr[self.rear] = val
        print(f"Added {val}")

    def dequeue(self):
        if self.is_empty():
            print("Empty Queue")
            return
        val = self.arr[self.front]
        if self.front >= self.rear:
            self.front = -1
            self.rear = -1
        else:
            self.front += 1
        print(f"Deleted {val}")

    def display(self):
        print("-----Queue-----")
        if self.is_empty():
            print("Empty")
            return
        print("vv Front")
        for i in range(self.front, self.rear + 1):
            print(f"{i} | {self.arr[i]}")
        print("^^ Rear")


if __name__ == "__main__":
    q = Queue()

    q.enqueue(2)
    q.display()
    q.enqueue(3)
    q.enqueue(5)
    q.display()
    q.dequeue()
    q.display()
    q.enqueue(7)
    q.enqueue(1)
    q.dequeue()
    q.display()
    q.enqueue(8)
    q.enqueue(0)
    q.display()

File: 08_StacksAndQueues/Queues/Queues CPP/Queues.cpp
#include "Queues.hpp"

Queue::Queue()
{
        front = -1;
        rear = -1;
        size = 7;
        arr = new int[size];
}

bool Queue::isFullQ()
{
        if(front == 0 && rear == size - 1)
                return true;
        return false;
}

bool Queue::isEmptyQ()
{
        if(front == -1)
                return true;
        return false;
}

void Queue::enQ(int val)
{
        if(isFullQ())
        {
                cout << "Queue is full\n";
                return;
        }
        if(front == -1)
                front = 0;
        rear++;
        arr[rear] = val;
        cout << "Added " << val << endl;
}

void Queue::deQ()
{
        int val;
        if(isEmptyQ())
        {
                cout << "Empty Queue\n";
                return;
        }
        val = arr[front];
        if(front >= rear)
        {
                front = -1;
                rear = -1;
        }
        else
        {
                front++;
        }

        cout << "Deleted " << val << endl;
} 

void Queue::displayQ()
{
        cout << "-----Queue-----\n";
        if(isEmptyQ())
        {
                cout << "Empty\n";
                return;
        }
        cout << "vv Front\n";
        for(int i = front; i <= rear; ++i)
        {

                cout << i << " | " << arr[i] << "\n";
        }
        cout << "^^ Rear" << endl;
}

int main()
{
        Queue q;

        q.enQ(2);
        q.displayQ();
        q.enQ(3); 
        q.enQ(5);
        q.displayQ();
        q.deQ();
        q.displayQ();
        q.enQ(7);
        q.enQ(1);
        q.deQ();
        q.displayQ();
        q.enQ(8);
        q.enQ(0);
        q.displayQ();

        return 0;
}